06week/higherOrder.js

```diff
@@ -3,27 +3,46 @@
 const assert = require('assert');
 
 function forEach(arr, callback) {
-  // Your code here
+  for(let i = 0; i < arr.length; i++) {
+    callback(arr[i]);
+  }
 }
 
 function map(arr, callback) {
-  // Your code here
+  let newArray = [];
+  for (let i = 0; i < arr.length; i++) {
+    let element = callback(arr[i]);
+    newArray.push(element);
+  }
+  return newArray;
 }
 
-function filter(arr, callback) {
-  // Your code here
-}
-
-function some(arr, callback) {
-  // Your code here
+function reduce(arr, callback,  acc) {
+  acc = acc || 0;
+  for (let i = 0; i < arr.length; i++) {
+    if (typeof arr[i] == 'number') {
+      acc = acc + arr[i];
+    } else if (typeof arr[i] == 'object') {
+      for (let j in arr[i]) {
+        acc = acc + arr[i][j];
+      }
+    }
+  }
+  return acc;
 }
 
-function every(arr, callback) {
-  // Your code here
+function filter(arr, callback) {
+  let newArray = [];
+  for (let i = 0; i < arr.length; i++) {
+    let element = callback(arr[i]);
+    if (element === true) {
```

Comment:
you should be able to just do 
` if (element) {`
because it gets coerced into a boolean when its put inside the if statement's conditional